synapse/storage/schema/main/delta/78/04_read_write_locks_triggers.sql.postgres

```diff
@@ -0,0 +1,75 @@
+/* Copyright 2023 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+-- c.f. 03_read_write_locks.sql for context.
+
+-- Add a foreign key constraint to ensure that if a lock is in
+-- `worker_read_write_locks_mode` then there must be a corresponding row in
+-- `worker_read_write_locks` (i.e. we don't accidentally end up with a row in
+-- `worker_read_write_locks_mode` when the lock is not currently acquired).
+ALTER TABLE worker_read_write_locks_mode ADD CONSTRAINT worker_read_write_locks_mode_foreign
+    FOREIGN KEY (lock_name, lock_key, token) REFERENCES worker_read_write_locks(lock_name, lock_key, token) DEFERRABLE INITIALLY DEFERRED;
```

Comment:
We discussed this a bit this morning that the foreign key has to be added after the fact because the `worker_read_write_locks` table isn't created yet when we first define the `worker_read_write_locks_mode` table.

Could we instead move this `ALTER TABLE` statement after `worker_read_write_locks` is created in `synapse/storage/schema/main/delta/78/03_read_write_locks.sql`

Reply:
> Could we instead move this `ALTER TABLE` statement after `worker_read_write_locks` is created in `synapse/storage/schema/main/delta/78/03_read_write_locks.sql`. Or maybe not because each delta is done in a single transaction?

We cannot, as we can't apply this to SQLite (due to them not support `ALTER TABLE`, and thus not supporting "circular" foreign key constraints). I'll add a comment.

Reply:
I wonder if you can temporarily disable foreign key constraints to get around the problem of not being able to add circular ones.

See section 7 of https://www.sqlite.org/lang_altertable.html: when doing other kinds of table recreations, they suggest you to disable the FKs temporarily, presumably to give you a window in which you can mess around with them and not get told off.

Reply:
Ok, so turns out it *is* possible to do foreign key constraints in SQLite, and it works out of the box if you just create the two tables in a transaction. However, that doesn't work for postgres, which means its non-trivial to support both.

Options here are:

1. Just have separate schema files for both SQLite and Postgres, making it non-obvious they're the same schema really.
2. Ignore the second foreign key constraint for SQLite, as we care less about it (and the triggers should ensure that the correct things happen anyway).

Reply:
Either way seems fine.

They're already pretty disparate with the trigger code so separating the schemas isn't a stretch.

And the other way is also fine since we were willing to not to have the foreign key in SQLite before we realized circular foreign keys was actually supported. My only thought around this is if we think the triggers should do the right thing, then perhaps we don't need the extra foreign key for Postgres either. But it's also just a easy thing to add to ensure the triggers are doing the right thing into the future :shrug:

Reply:
I've opted to split the schemas.